File: src/main/java/org/example/kalkulationsprogramm/domain/KasseEinstellung.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Singleton-Konfiguration des Kassenbuchs (max. 1 Zeile).
 *
 * Steuert:
 *   - Kassen-Mindestbestand: BelegService verhindert Buchungen, die den Saldo
 *     unter diesen Wert fallen lassen wuerden. Default 0 EUR.
 *   - Ehegattengehalt-Automatik: Monatlicher Scheduler bucht am konfigurierten
 *     Tag den Betrag als Kassenausgabe. Wenn der Bar-Saldo nicht reicht, wird
 *     vorher automatisch eine Privateinlage in genau passender Hoehe gebucht.
 *
 * letzteBuchungJahrmonat ist die Idempotenz-Sperre: pro YYYY-MM nur einmal
 * buchen, auch wenn der Scheduler mehrfach laeuft.
 */
@Getter
@Setter
@Entity
@Table(name = "kasse_einstellung")
public class KasseEinstellung {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal mindestbestand = BigDecimal.ZERO;

    @Column(name = "ehegattengehalt_aktiv", nullable = false)
    private boolean ehegattengehaltAktiv = false;

    @Column(name = "ehegattengehalt_betrag", precision = 10, scale = 2)
    private BigDecimal ehegattengehaltBetrag;

    /**
     * Tag des Monats (1-28). Max. 28, damit jeder Monat denselben Stichtag
     * haben kann.
     */
    @Column(name = "ehegattengehalt_tag")
    private Integer ehegattengehaltTag;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ehegattengehalt_sachkonto_id")
    private Sachkonto ehegattengehaltSachkonto;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ehegattengehalt_kostenstelle_id")
    private Kostenstelle ehegattengehaltKostenstelle;

    @Column(name = "ehegattengehalt_empfaenger_name", length = 120)
    private String ehegattengehaltEmpfaengerName;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "privateinlage_sachkonto_id")
    private Sachkonto privateinlageSachkonto;

    @Column(name = "letzte_buchung_jahrmonat", length = 7)
    private String letzteBuchungJahrmonat;

    @Column(name = "aktualisiert_am")
    private LocalDateTime aktualisiertAm;

    @PreUpdate
    @PrePersist
    void onSave() {
        aktualisiertAm = LocalDateTime.now();
        if (mindestbestand == null) {
            mindestbestand = BigDecimal.ZERO;
        }
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/KasseEinstellung.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "kasse_einstellung")
open class KasseEinstellung {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @Column(nullable = false, precision = 10, scale = 2)
    open var mindestbestand: BigDecimal? = BigDecimal.ZERO

    @Column(name = "ehegattengehalt_aktiv", nullable = false)
    open var ehegattengehaltAktiv: Boolean = false

    @Column(name = "ehegattengehalt_betrag", precision = 10, scale = 2)
    open var ehegattengehaltBetrag: BigDecimal? = null

    @Column(name = "ehegattengehalt_tag")
    open var ehegattengehaltTag: Int? = null

    @Column(name = "ehegattengehalt_empfaenger_name", length = 120)
    open var ehegattengehaltEmpfaengerName: String? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "privateinlage_sachkonto_id")
    open var privateinlageSachkonto: Sachkonto? = null

    @Column(name = "letzte_buchung_jahrmonat", length = 7)
    open var letzteBuchungJahrmonat: String? = null

    @Column(name = "aktualisiert_am")
    open var aktualisiertAm: LocalDateTime? = null

    open fun isEhegattengehaltAktiv(): Boolean = ehegattengehaltAktiv

}
